src/k8s/pkg/client/kubernetes/upgrades.go

```diff
@@ -0,0 +1,178 @@
+	UpgradePhaseFeatureUpgrade = "FeatureUpgrade"
+	UpgradePhaseFailed         = "Failed"
+	UpgradePhaseCompleted      = "Completed"
+)
+
+const (
+	kind    = "Upgrade"
+	group   = "k8sd.io"
+	version = "v1alpha"
+)
+
+type UpgradeMetadata struct {
+	Name string `json:"name,omitempty"`
+}
+
+type UpgradeStatus struct {
+	Phase         string   `json:"phase,omitempty"`
+	UpgradedNodes []string `json:"upgradedNodes,omitempty"`
+}
+
+type Upgrade struct {
+	APIVersion string          `json:"apiVersion,omitempty"`
+	Kind       string          `json:"kind,omitempty"`
+	Metadata   UpgradeMetadata `json:"metadata,omitempty"`
+	Status     UpgradeStatus   `json:"status,omitempty"`
+}
+
+func NewUpgrade(name string) Upgrade {
+	return Upgrade{
+		APIVersion: group + "/" + version,
+		Kind:       kind,
+		Metadata:   UpgradeMetadata{Name: name},
+		Status:     UpgradeStatus{Phase: UpgradePhaseNodeUpgrade, UpgradedNodes: []string{}},
+	}
+}
+
+func (c *Client) k8sdIoRestClient() (*rest.RESTClient, error) {
+	k8sdConfig := c.RESTConfig()
+	k8sdConfig.GroupVersion = &schema.GroupVersion{Group: group, Version: version}
+	k8sdConfig.NegotiatedSerializer = serializer.NewCodecFactory(runtime.NewScheme())
+
+	restClient, err := rest.RESTClientFor(k8sdConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create REST client for k8sd.io group: %w", err)
+	}
+	return restClient, nil
+}
+
+// GetInProgressUpgrade returns the upgrade CR that is currently in progress.
+// TODO(ben): Maybe make this more generic, e.g. GetUpgrade(filterFunc func(Upgrade) bool) (*Upgrade, error)
+func (c *Client) GetInProgressUpgrade(ctx context.Context) (*Upgrade, error) {
+	log := log.FromContext(ctx).WithValues("upgrades", "GetInProgressUpgrade")
+
+	restClient, err := c.k8sdIoRestClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create REST client for k8sd.io group: %w", err)
+	}
+
+	upgrades, err := restClient.Get().AbsPath(fmt.Sprintf("/apis/%s/%s/upgrades", group, version)).DoRaw(ctx)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// No upgrade in progress.
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to get upgrades: %w", err)
+	}
+
+	var result struct {
+		Items []Upgrade `json:"items"`
+	}
+	if err := json.Unmarshal(upgrades, &result); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal upgrades: %w", err)
+	}
+
+	var matches []Upgrade
+	for _, upgrade := range result.Items {
+		if upgrade.Status.Phase != UpgradePhaseFailed && upgrade.Status.Phase != UpgradePhaseCompleted {
+			matches = append(matches, upgrade)
+		}
+	}
+	if len(matches) == 0 {
+		return nil, nil
+	}
+	if len(matches) > 1 {
+		log.Info("Warning: Found multiple in-progress upgrades", "inprogress upgrades", len(matches))
+	}
+	// Sort matches by name
+	sort.Slice(matches, func(i, j int) bool {
+		return matches[i].Metadata.Name < matches[j].Metadata.Name
+	})
+
+	// Return the latest
+	return &matches[len(matches)-1], nil
```

Comment:
nit: Can we just assign `len(matches)` to a variable and reuse instead of recalculating every time?